### PR TITLE
Feature/validation base system trng 1219

### PR DESCRIPTION
**Pull request by SimonTheSourcerer**
### Description
New base validation system, this is without proper implementation useless. The implementation will be added via creator pro.

**Fixes**: # (issue)

**ToDo:**
- [x] Documentation has to be integrated somehow

### Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**Test Configuration**:
Does need additional branches:
https://github.com/Innoactive/Creator-Pro/pull/2
https://github.com/Innoactive/Basic-Interaction-Component/pull/33
https://github.com/Innoactive/Basic-Conditions-And-Behaviors/pull/29

**Setup**
Copy and paste from Assets folder for a fresh clone 

```
git clone --single-branch --branch feature/validation-trng-374 git@github.com:Innoactive/Basic-Conditions-And-Behaviors.git Innoactive/Creator/Components/Basic-ConditionsBehaviors
git clone --single-branch --branch feature/validation-ui-trng-1098 git@github.com:Innoactive/Basic-Interaction-Component.git Innoactive/Creator/Components/Basic-Interaction
git clone --single-branch --branch develop git@github.com:Innoactive/TextToSpeech-Component.git Innoactive/Creator/Components/TextToSpeech
git clone --single-branch --branch develop git@github.com:Innoactive/XR-Interaction-Component.git Innoactive/Creator/Components/XR-Interaction-Component
git clone --single-branch --branch feature/validation-base-system-trng-1219 git@github.com:Innoactive/Creator.git Innoactive/Creator/Core
git clone --single-branch --branch develop git@github.com:Innoactive/IA-Training-Template.git Innoactive/Creator/IA-Training-Template
git clone --single-branch --branch feature/validation-system git@github.com:Innoactive/Creator-Pro.git Innoactive/Creator/Pro
```

### CODE REVIEW

**Comment by Luphoxa** [comment]
Questions
1. How are empty steps handled? If I have an empty step in the course, the chapter shows a warning sign, but the step doesn't
2. I am not getting the "step is not reachable" warning anymore. Was it taken out on purpose?

**Comment by SimonTheSourcerer** [comment]
Empty steps are not handled.
Not reachable steps should work (again), pushed a fix for that today.

**Comment by Luphoxa** [comment]
Okay, I will test it next week again, at latest on thursday 👍

File: Editor/Configuration/DefaultEditorConfiguration.cs [hunk 1]
```diff
@@ -59,5 +60,16 @@ public virtual AllowedMenuItemsSettings AllowedMenuItemsSettings
...
+
+        internal virtual IValidationHandler Validation { get; }
+
+        protected DefaultEditorConfiguration()
+        {
+#if CREATOR_PRO
+            Validation = new ValidationHandler();
+#else
+            Validation = new NoValidationHandler();
+#endif
```
Comment (Gusinuhe) [inline Editor/Configuration/DefaultEditorConfiguration.cs:72]: There is a problem: 

If `Innoactive.CreatorEditor.CourseValidation.ValidationHandler` is missing and `CREATOR_PRO` is already defined in the project, `AssemblySymbolChecker` will not even compile to remove the `CREATOR_PRO` symbol.

What if you make `ValidationHandler` inherit from `NoValidationHandler` and then you simply look for (with reflection) the deepest implementation of `IValidationHandler`? 🤔

File: Editor/CourseValidation/IValidationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Collections.Generic;
+using Innoactive.Creator.Core;
+
+namespace Innoactive.CreatorEditor.CourseValidation
+{
+    /// <summary>
+    /// ValidationHandler handles the validation of course the context.
```
Comment (Gusinuhe) [inline Editor/CourseValidation/IValidationHandler.cs:7]: This description is very odd 🤔
Sorry, I am kind of block now and I can not thing of anything better to suggest for now 😕

File: Editor/CourseValidation/IValidationReport.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
+﻿using System.Collections.Generic;
+using System.Reflection;
+using Innoactive.Creator.Core;
+using Innoactive.Creator.Core.Behaviors;
+using Innoactive.Creator.Core.Conditions;
+
+namespace Innoactive.CreatorEditor.CourseValidation
+{
+    /// <summary>
+    /// Will be generated by the <see cref="IValidationHandler"/>.
```
Comment (Gusinuhe) [inline Editor/CourseValidation/IValidationReport.cs:10]: What? 🤔

File: Editor/CourseValidation/IValidationReport.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        /// </summary>
+        List<EditorReportEntry> GetContextEntriesFor<T>(IStepData step) where T : IContext;
+
+        /// <summary>
+        /// Returns all <see cref="EditorReportEntry"/> found for given step.
+        /// </summary>
+        List<EditorReportEntry> GetEntriesFor(IData data, MemberInfo info);
+
+        /// <summary>
+        /// Get Entries for IBehaviorData.
```
Comment (Gusinuhe) [inline Editor/CourseValidation/IValidationReport.cs:50]:
```suggestion
        /// Get Entries for <see cref="IBehaviorData"/>.
```

File: Editor/CourseValidation/IValidationReport.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        /// </summary>
+        List<EditorReportEntry> GetEntriesFor(IBehaviorData data);
+
+        /// <summary>
+        /// Get Entries for IConditionData
```
Comment (Gusinuhe) [inline Editor/CourseValidation/IValidationReport.cs:55]:
```suggestion
        /// Get Entries for <see cref="IConditionData"/>
```

File: Editor/CourseValidation/NoValidationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+﻿using Innoactive.Creator.Core;
+using Innoactive.CreatorEditor.CourseValidation;
+
+/// <summary>
+/// Used to disable validation overall.
+/// </summary>
+internal class NoValidationHandler : IValidationHandler
```
Comment (Gusinuhe) [inline Editor/CourseValidation/NoValidationHandler.cs:7]: Perhaps, `DisabledValidationHandler` 🤔

File: Editor/CreatorProjectSettings.cs [hunk 7]
```diff
@@ -13,6 +13,18 @@ public class CreatorProjectSettings : ScriptableObject
...
+    /// Enables the validation system.
+    /// </summary>
+    [SerializeField]
+    public bool IsValidationEnabled = true;
+
+    /// <summary>
+    /// Enables the validation system debugging output.
+    /// </summary>
+    [SerializeField]
+    public bool DebugValidationSystem = false;
```
Comment (Gusinuhe) [inline Editor/CreatorProjectSettings.cs:26]: Should this be in the core? It is pretty useless without the validation thingi.

File: Editor/DefaultEditingStrategy.cs [hunk 8]
```diff
@@ -12,7 +15,8 @@ internal class DefaultEditingStrategy : IEditingStrategy
     {
         private CourseWindow courseWindow;
         private StepWindow stepWindow;
-        private ICourse course;
+
+        public ICourse CurrentCourse { get; protected set; }
```
Comment (Gusinuhe) [inline Editor/DefaultEditingStrategy.cs:19]: Missing docs

File: Editor/TestTools/EditorImguiTester/EmptyTestStrategy.cs [hunk 9]
```diff
@@ -8,6 +8,8 @@ namespace Innoactive.CreatorEditor.TestTools
     /// </summary>
     internal class EmptyTestStrategy : IEditingStrategy
     {
+        public ICourse CurrentCourse { get; }
```
Comment (Gusinuhe) [inline Editor/TestTools/EditorImguiTester/EmptyTestStrategy.cs:11]: Missing docs 😇

File: Editor/CourseValidation/NoValidationHandler.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+    public bool IsAllowedToValidate()
+    {
+        return false;
+    }
+
+    public IValidationReport Validate(IData data, ICourse course, IContext context = null)
+    {
+        return null;
+    }
```
Comment (Gusinuhe) [inline Editor/CourseValidation/NoValidationHandler.cs:21]: Missing docs

File: Editor/UI/Graphics/StepNode.cs [hunk 11]
```diff
@@ -14,13 +15,16 @@ internal class StepNode : EditorNode
...
         private bool isLastSelectedStep;
+
         private IStep step;
 
         //The last step that was clicked on is selected.
         private readonly GraphicalElementRenderer renderer;
 
         public CreateTransitionButton CreateTransitionButton { get; private set; }
 
+        public IChapter ActiveChapter { get; }
```
Comment (Gusinuhe) [inline Editor/UI/Graphics/StepNode.cs:26]: Missing docs

**Review by Gusinuhe — CHANGES_REQUESTED** [review]
There is an error in `.Documentation/articles/toc.yml` that prevents the articles section to be rendered:

> [20-10-26 01:08:45.018]Error:[BuildCommand.BuildCore.Build Document.Load.TocDocumentProcessor](articles/toc.yml)/Users/gustavoquiroz/Desktop/Validation System/Assets/Innoactive/Creator/Core/.Documentation/articles/toc.yml is not a valid TOC File: /Users/gustavoquiroz/Desktop/Validation System/Assets/Innoactive/Creator/Core/.Documentation/articles/toc.yml is not a valid TOC file, detail: (Line: 88, Col: 11, Idx: 2966) - (Line: 88, Col: 11, Idx: 2966): Mapping values are not allowed in this context..

File: .Documentation/articles/pro/01-validation.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
+# Validation System
+
+The Validation System is part of our PRO offer. It allows you to easily find errors in your training courses. The system allows you to add new validations and also use the existing once for your own implemented Behaviors and Conditions. 
+
+**What do have to take into account to get validation properly working?**
+
+- The validation system can be enabled/disabled by switchting the CreatorProjectSettings entry `IsValidationEnabled`.
+- Validation only runs when a `RuntimeConfigurator` is existent in the scene.
+- All fields with the attribute `[DateMember]` in Behaviors and Conditions will be taken as required, if this is not the case you have add a `[OptionalValue]` attribute.
```
Comment (tomwim) [inline .Documentation/articles/pro/01-validation.md:9]:
```suggestion
- All fields with the attribute `[DateMember]` in Behaviors and Conditions will be taken as required, if this is not the case you have to add a `[OptionalValue]` attribute.
```

File: .Documentation/articles/pro/01-validation.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+Overall validation is done by two different validator types, BaseValidation is scope specific for a `IData` objects, while the AttributeValidator is assigned to a field or property in any `BehaviorData` or `ConditionData`. Validation is run over the whole course data structure which means the BaseValidator can be run for `CourseData, ChapterData, StepData, BehaviorData, ConditionData`
```
Comment (tomwim) [inline .Documentation/articles/pro/01-validation.md:11]:
```suggestion
Overall validation is done by two different validator types. BaseValidation is scope specific for a `IData` objects, while the AttributeValidator is assigned to a field or property in any `BehaviorData` or `ConditionData`. Validation is run over the whole course data structure which means the BaseValidator can be run for `CourseData, ChapterData, StepData, BehaviorData, ConditionData`
```

File: .Documentation/articles/pro/01-validation.md [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+## Validation via BaseValidator
+
+To implement a new Validator you have to extend `BaseValidator<T, TContext>` or `BaseStepValidator`. Newly implemented Validator for Chapter- and StepContext will be automatically found and used while validating the specific Context.
```
Comment (tomwim) [inline .Documentation/articles/pro/01-validation.md:15]:
```suggestion
To implement a new Validator you have to extend `BaseValidator<T, TContext>` or `BaseStepValidator`. Newly implemented Validators for Chapter- and StepContext will be automatically found and used while validating the specific Context.
```

File: .Documentation/articles/pro/01-validation.md [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+* **T** is the target type which will be validated, usually the Data objects IChapterData or IStepData are used here.
+* **TContext** is the Context the target Type T will be validated in, which has to be the fitting Context for the used type, for example StepContext for IStepData.
+
+One example for an ChapterContext validator is the [StepConnectionValidator](https://github.com/Innoactive/Creator/blob/develop/Editor/CourseValidation/Validator/StepConnectionValidator.cs)
```
Comment (tomwim) [inline .Documentation/articles/pro/01-validation.md:20]:
```suggestion
One example for a ChapterContext validator is the [StepConnectionValidator](https://github.com/Innoactive/Creator/blob/develop/Editor/CourseValidation/Validator/StepConnectionValidator.cs)
```

File: .Documentation/articles/pro/01-validation.md [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+## Validation via Attribute Validation
+
+For validating Behaviors and Condition Data we implemented the support of attribute validators, which can be used to validate specific fields.
+
+The `StepAttributeValidator` checks for all Behaviors & Condition if there are any Attributes added which implement the interface `IAttributeValidator`. All found validators will be run and the reports given back to the ValidationHandler.
```
Comment (tomwim) [inline .Documentation/articles/pro/01-validation.md:26]:
```suggestion
The `StepAttributeValidator` checks for all Behaviors & Conditions if there are any Attributes added which implement the interface `IAttributeValidator`. All found validators will be run and the reports given back to the ValidationHandler.
```

File: .Documentation/articles/pro/01-validation.md [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+To implement your own AttributeValidation you have to implement `IAttributeValidator`.
+
+An Example AttributeValidator can be found at Github: [CheckForComponentAttribute](https://github.com/Innoactive/Creator/blob/develop/Runtime/CourseValidation/CheckForComponentAttribute.cs)
+
+Hint: *`AttributeValidators` have to be placed into runtime assemblies due to beeing attached to course structure.*
```
Comment (tomwim) [inline .Documentation/articles/pro/01-validation.md:31]:
```suggestion
Hint: *`AttributeValidators` have to be placed into runtime assemblies due to being attached to course structure.*
```
🐝 🐝 🐝

File: .Documentation/articles/pro/index.md [hunk 18]
```diff
@@ -0,0 +1,7 @@
+# Overview
```
Comment (tomwim) [inline .Documentation/articles/pro/index.md:1]: Is this file needed? How do you get to articles/pro/index ? Isn't pro more an extension of core?

File: Editor/CourseValidation/IValidationHandler.cs (hunk 2 above)
Comment (tomwim) [reply Editor/CourseValidation/IValidationHandler.cs:7]: `ValidationHandler validates data objects, e.g. steps or conditions, of a given course and reports whether the object causes issues or is faulty, e.g. empty fields or invalid values in a behavior.`

File: Editor/CreatorProjectSettings.cs (hunk 7 above)
Comment (tomwim) [reply Editor/CreatorProjectSettings.cs:26]: Either under a "Pro" section or only visible if you have Pro?

**Review by tomwim — CHANGES_REQUESTED** [review]

**Comment by tomwim** [comment]
Code reviewed, will check functionality asap

File: Editor/DefaultEditingStrategy.cs (hunk 8 above)
Comment (SimonTheSourcerer) [reply Editor/DefaultEditingStrategy.cs:19]: Still an internal class

File: Editor/TestTools/EditorImguiTester/EmptyTestStrategy.cs (hunk 9 above)
Comment (SimonTheSourcerer) [reply Editor/TestTools/EditorImguiTester/EmptyTestStrategy.cs:11]: Still an internal class

File: Editor/CourseValidation/NoValidationHandler.cs (hunk 10 above)
Comment (SimonTheSourcerer) [reply Editor/CourseValidation/NoValidationHandler.cs:21]: Still an internal class

File: Editor/UI/Graphics/StepNode.cs (hunk 11 above)
Comment (SimonTheSourcerer) [reply Editor/UI/Graphics/StepNode.cs:26]: Still an internal class

File: Editor/Configuration/DefaultEditorConfiguration.cs (hunk 1 above)
Comment (SimonTheSourcerer) [reply Editor/Configuration/DefaultEditorConfiguration.cs:72]: Makes it very prone to breaking if there are at any time more than one type of ValidationHandler. But I cannot come up with a better idea, I think about it.

File: Editor/CourseValidation/DisabledValidationHandler.cs [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+using Innoactive.CreatorEditor.CourseValidation;
+
+/// <summary>
+/// Does not validate, used to disabled the validation system.
+/// </summary>
+internal class DisabledValidationHandler : IValidationHandler
+{
+    public IContextResolver ContextResolver { get; set; } = null;
+
+    public IValidationReport LastReport { get; } = null;
```
Comment (Gusinuhe) [inline Editor/CourseValidation/DisabledValidationHandler.cs:11]: Missing docs 😇

File: Editor/CreatorProjectSettings.cs (hunk 7 above)
Comment (SimonTheSourcerer) [reply Editor/CreatorProjectSettings.cs]: Made the class partial, which allows different components to contribute settings, but I'm not sure this is a good way. If stuff is missing and it's changed/saved parts of settings are lost.

File: Editor/CourseValidation/DisabledValidationHandler.cs (hunk 19 above)
Comment (SimonTheSourcerer) [reply Editor/CourseValidation/DisabledValidationHandler.cs:11]: still internal class...

File: .Documentation/articles/pro/index.md (hunk 18 above)
Comment (SimonTheSourcerer) [reply .Documentation/articles/pro/index.md:1]: rewrote it, forgot to "properly" fill it. Dont know if we should keep it that way. Are there any better ideas how to place the CreatorPro stuff in the documentation?

Comment (Gusinuhe) [reply .Documentation/articles/pro/index.md:1]: @tomwim there is a new section in the articles called `Creator Pro` or something like that.

**Review by Gusinuhe — CHANGES_REQUESTED** [review]
After fixing a warning in a condition, the warning icon disappear from the condition but not from the transition:

**After:**
Screenshot 2020-10-27 at 12 39 36

**Before**
Screenshot 2020-10-27 at 12 40 22

**Comment by SimonTheSourcerer** [comment]
^ fixed

**Review by Gusinuhe — CHANGES_REQUESTED** [review]
Another error similar to the old one:

Giving a step that is floating around without being connected, a warning is shown in the chapter and in the step:

image

After removing the step, the warning continues in the chapter:

image

I also tested to connect the step, the warning in the step disappears but the warning in the chapter continues:

image

**Comment by SimonTheSourcerer** [comment]
Works perfectly for me. Can you show me the steps to reproduce it?

**Comment by Gusinuhe** [comment]
Validation bug

**Comment by SimonTheSourcerer** [comment]
Can anyone reproduce this? I cant :(

**Comment by Gusinuhe** [comment]
**Given** a course with some warnings opened in its corresponding training scene.
**When** switching to a new scene (or a scene that is not set up as a training scene).
**Then** the validation system stops working, however, the course can still be edited and saved.

2020-10-28 at 8 59 25 - Indigo Macaw 3

**Comment by Gusinuhe** [comment]
> Can anyone reproduce this? I cant :(

Whoops! False alarm! :innocent:

It does work correctly, the issue was that there was another step super hidden that was also not connected to the workflow.

**Review by Gusinuhe — APPROVED** [review]
Since the last bug is supposed to be addressed separately I got no more objection for this PR, auto-approval! ✔️

File: Editor/Configuration/DefaultEditorConfiguration.cs (hunk 1 above)
Comment (SimonTheSourcerer) [reply Editor/Configuration/DefaultEditorConfiguration.cs]: No better idea found, as far we are going to accept the fact that going back from pro does need more than just deleting the folder.

**Review by tomwim — APPROVED** [review]

**Review by Maxsiom — APPROVED** [review]
